transports/quic/src/transport.rs

```diff
@@ -191,18 +214,65 @@ impl<P: Provider> Transport for GenTransport<P> {
                 &mut listeners[index].dialer_state
             }
         };
-        Ok(dialer_state.new_dial(socket_addr, self.handshake_timeout, version))
+        Ok(dialer_state.new_dial(socket_addr, handshake_timeout, version))
     }
 
     fn dial_as_listener(
         &mut self,
         addr: Multiaddr,
     ) -> Result<Self::Dial, TransportError<Self::Error>> {
-        // TODO: As the listener of a QUIC hole punch, we need to send a random UDP packet to the
-        // `addr`. See DCUtR specification below.
-        //
-        // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
-        Err(TransportError::MultiaddrNotSupported(addr))
+        let (socket_addr, _version, peer_id) = self.remote_multiaddr_to_socketaddr(addr.clone())?;
+
+        let peer_id = peer_id.ok_or(TransportError::MultiaddrNotSupported(addr))?;
+
+        let listeners = self.eligible_listeners(&socket_addr);
+
+        let endpoint_channel = match listeners.len() {
+            0 => {
+                return Err(TransportError::Other(
+                    Error::NoActiveListenerForDialAsListener,
+                ));
+            }
+            1 => listeners[0].endpoint_channel.clone(),
+            _ => {
+                // Pick any listener to use for dialing.
+                // We hash the socket address to achieve determinism.
+                let mut hasher = DefaultHasher::new();
+                socket_addr.hash(&mut hasher);
+                let index = hasher.finish() as usize % listeners.len();
+                listeners[index].endpoint_channel.clone()
+            }
+        };
+
+        let hole_puncher = HolePuncher::new(endpoint_channel, socket_addr, self.handshake_timeout);
+
+        let (sender, receiver) = oneshot::channel();
+
+        match self
+            .hole_punch_map
+            .lock()
+            .unwrap()
+            .entry((socket_addr, peer_id))
+        {
+            Entry::Vacant(entry) => entry.insert(sender),
+            Entry::Occupied(_) => {
+                return Err(TransportError::Other(Error::HolePunchInProgress(
+                    socket_addr,
+                    peer_id,
+                )));
+            }
+        };
+
+        Ok(Box::pin(async {
+            futures::select! {
+                hole_punched = receiver.fuse() => {
+                    Ok(hole_punched.unwrap())
+                },
+                err = hole_puncher.fuse() => {
+                    Err(err)
+                }
+            }
```

Comment:
I would prefer using [`futures::future::select`](https://docs.rs/futures/latest/futures/future/fn.select.html), thus not introducing the complexity of a macro and, as far as I can tell, dropping the need for these futures to be `Fuse`.